### PR TITLE
CSS 2021 queries

### PR DESCRIPTION
**Pull request by rviscomi**
Progress on #2140 

Over 100 queries needed for this chapter, most are a repeat of [2020 queries](https://github.com/HTTPArchive/almanac.httparchive.org/tree/main/sql/2020/css) so not listing them out.

- [x] 2020 queries ported to 2021
- [x] `:focus` customization, and prevalence of `outline: 0`
- [x] `:focus-visible` use
- [x] Collecting generated content strings (e.g., `content: "Item "`)
- [ ] Formats of images called in CSS (GIF/JPG/PNG/WEBM/SVG/gradients)
- [ ] Number of images per style sheet
- [ ] Weight of images called in CSS
- [ ] Pixel sizes of images called in CSS
- [ ] `animation-duration` and `transition-duration` values
- [ ] `animation-delay` and `transition-delay` values
- [ ] Counts and percentage positions of animation keyframes
- [x] CSS Ruby property usage
- [x] Usage of CSS utility frameworks
- [ ] Usage of polyfills
- [x] `@property` usage (% of pages using, median rules pp, stats about types, inherits, initial values)

Opening this PR for review despite not all 2021 metrics being implemented yet. I'll create a separate PR for those. All of the other 100+ queries besides [content_values.sql](https://github.com/HTTPArchive/almanac.httparchive.org/pull/2306/files#diff-c62ad5938ffaff2084bc5ad7ea01e071220cb8bdbd12756d1752945578892de6) and [focus_visible.sql](https://github.com/HTTPArchive/almanac.httparchive.org/pull/2306/files#diff-afae1fbc849377efd7aacbba0b11886fd67e204de566688bbbe014bb3a803f6c) are carry-overs from 2020, which probably don't need to be reviewed too closely since they were already reviewed for the most part. (and I can't imagine anyone wants to thoroughly review 100+ queries anyway 😆)

### CODE REVIEW

**Comment by rviscomi** [comment]
Copied all 2020 queries into the 2021 directory and ran this script to use the latest Almanac data:

```sh
sed -i "" -e "s/202008/202107/g" \
  -e "s/2020_08_01/2021_07_01/g" \
  -e "s/2020-08-01/2021-07-01/g" \
  sql/2021/css/*.sql
```

File: sql/2021/css/box_sizing.sql [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+FROM (
+  SELECT
+    client,
+    page,
+    countBorderBoxDeclarations(css) AS declarations
+  FROM
+    `httparchive.almanac.parsed_css`
+  WHERE
+    date = '2021-07-01'),
+  UNNEST([10, 25, 50, 75, 90]) AS percentile
```
Comment (tunetheweb) [inline sql/2021/css/box_sizing.sql:33]: Potential issue with UNNEST and count.

File: sql/2021/css/content_values.sql [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    WHERE
+      date = '2021-07-01'
+    GROUP BY
+      client)
+  JOIN
+    `httparchive.almanac.parsed_css`
+  USING
+    (client),
+    UNNEST(getContentStrings(css)) AS content
+  WHERE
```
Comment (tunetheweb) [inline sql/2021/css/content_values.sql:52]: Potential UNNEST count issue.

File: sql/2021/css/custom_property_registered.sql [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+SELECT DISTINCT
+  _TABLE_SUFFIX AS client,
+  prop,
+  COUNT(DISTINCT url) OVER (PARTITION BY _TABLE_SUFFIX, prop) AS pages,
+  COUNT(DISTINCT url) OVER (PARTITION BY _TABLE_SUFFIX) AS total,
+  COUNT(DISTINCT url) OVER (PARTITION BY _TABLE_SUFFIX, prop) / COUNT(DISTINCT url) OVER (PARTITION BY _TABLE_SUFFIX) AS pct
+FROM
+  `httparchive.pages.2021_07_01_*`,
+  UNNEST(getCustomPropertiesWithComputedStyle(payload)) AS prop
```
Comment (tunetheweb) [inline sql/2021/css/custom_property_registered.sql:51]: Potenital UNNEST counting issue

File: sql/2021/css/houdini_animated_custom_properties.sql [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  custom_property,
+  COUNT(DISTINCT page) OVER (PARTITION BY client, custom_property) AS pages,
+  COUNT(DISTINCT page) OVER (PARTITION BY client) AS total_pages,
+  COUNT(DISTINCT page) OVER (PARTITION BY client, custom_property) / COUNT(DISTINCT page) OVER (PARTITION BY client) AS pct_pages,
+  COUNT(0) OVER (PARTITION BY client, custom_property) AS freq,
+  COUNT(0) OVER (PARTITION BY client) AS total,
+  COUNT(0) OVER (PARTITION BY client, custom_property) / COUNT(0) OVER (PARTITION BY client) AS pct
+FROM
+  `httparchive.almanac.parsed_css`,
+  UNNEST(getAnimatedCustomProperties(css)) AS custom_property
```
Comment (tunetheweb) [inline sql/2021/css/houdini_animated_custom_properties.sql:38]: Potential UNNEST count issue.

File: sql/2021/css/calc_properties.sql [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    UNNEST(getCalcProperties(css)) AS prop
+  WHERE
+    date = '2021-07-01' AND
+    # Limit the size of the CSS to avoid OOM crashes.
+    LENGTH(css) < 0.1 * 1024 * 1024)
+GROUP BY
+  client,
+  name
+HAVING
+  pages >= 100
```
Comment (tunetheweb) [inline sql/2021/css/calc_properties.sql:91]: HAVING will affect PARTITION counts.

File: sql/2021/css/calc_units.sql [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    UNNEST(getCalcUnits(css)) AS unit
+  WHERE
+    date = '2021-07-01' AND
+    # Limit the size of the CSS to avoid OOM crashes.
+    LENGTH(css) < 0.1 * 1024 * 1024)
+GROUP BY
+  client,
+  unit
+HAVING
+  pages >= 100
```
Comment (tunetheweb) [inline sql/2021/css/calc_units.sql:91]: HAVING will affect PARTITION COUNTS

File: sql/2021/css/all_properties.sql [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+  FROM
+    `httparchive.almanac.parsed_css`,
+    UNNEST(getProperties(css)) AS prop
+  WHERE
+    date = '2021-07-01'
+  GROUP BY
+    client,
+    prop)
+WHERE
+  pages >= 1000
```
Comment (tunetheweb) [inline sql/2021/css/all_properties.sql:52]: Will affect PARTITION counts

File: sql/2021/css/sass_combined_variable_names.sql [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+FROM (
+  SELECT
+    _TABLE_SUFFIX AS client,
+    name,
+    COUNT(0) AS freq,
+    SUM(COUNT(0)) OVER (PARTITION BY _TABLE_SUFFIX) AS total,
+    COUNT(0) / SUM(COUNT(0)) OVER (PARTITION BY _TABLE_SUFFIX) AS pct
+  FROM
+    `httparchive.pages.2021_07_01_*`,
+    UNNEST(getCombinedVariableNames(payload)) AS name
```
Comment (tunetheweb) [inline sql/2021/css/sass_combined_variable_names.sql:28]: Potential UNNEST total issue

File: sql/2021/css/sourcemap_preprocessors.sql [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+SELECT
+  _TABLE_SUFFIX AS client,
+  ext,
+  COUNT(0) AS freq,
+  SUM(COUNT(0)) OVER (PARTITION BY _TABLE_SUFFIX) AS total,
+  COUNT(0) / SUM(COUNT(0)) OVER (PARTITION BY _TABLE_SUFFIX) AS pct
+FROM
+  `httparchive.pages.2021_07_01_*`,
+  UNNEST(getSourcemappedExts(payload)) AS ext
```
Comment (tunetheweb) [inline sql/2021/css/sourcemap_preprocessors.sql:21]: Potential total issue

File: sql/2021/css/supports_criteria.sql [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+SELECT
+  client,
+  supports,
+  COUNT(DISTINCT page) AS pages,
+  COUNT(0) AS freq,
+  SUM(COUNT(0)) OVER (PARTITION BY client) AS total,
+  COUNT(0) / SUM(COUNT(0)) OVER (PARTITION BY client) AS pct
+FROM
+  `httparchive.almanac.parsed_css`,
+  UNNEST(getSupports(css)) AS supports
```
Comment (tunetheweb) [inline sql/2021/css/supports_criteria.sql:58]: Potential counting issue

File: sql/2021/css/units_zero.sql [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+  SELECT
+    client,
+    unit.unit,
+    COUNT(DISTINCT page) AS pages,
+    SUM(unit.freq) AS freq,
+    SUM(SUM(unit.freq)) OVER (PARTITION BY client) AS total,
+    SUM(unit.freq) / SUM(SUM(unit.freq)) OVER (PARTITION BY client) AS pct
+  FROM
+    `httparchive.almanac.parsed_css`,
+    UNNEST(getZeroUnits(css)) AS unit
```
Comment (tunetheweb) [inline sql/2021/css/units_zero.sql:107]: Potenitally incorrect TOTAL

File: sql/2021/css/vendor_prefix_functions.sql [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+  SELECT
+    client,
+    function,
+    COUNT(DISTINCT page) AS pages,
+    COUNT(0) AS freq,
+    SUM(COUNT(0)) OVER (PARTITION BY client) AS total,
+    COUNT(0) / SUM(COUNT(0)) OVER (PARTITION BY client) AS pct
+  FROM
+    `httparchive.almanac.parsed_css`,
+    UNNEST(getVendorPrefixFunctions(css)) AS function
```
Comment (tunetheweb) [inline sql/2021/css/vendor_prefix_functions.sql:95]: Potenital incorrect total

File: sql/2021/css/vendor_prefix_keywords.sql [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+  SELECT
+    client,
+    keyword,
+    COUNT(DISTINCT page) AS pages,
+    COUNT(0) AS freq,
+    SUM(COUNT(0)) OVER (PARTITION BY client) AS total,
+    COUNT(0) / SUM(COUNT(0)) OVER (PARTITION BY client) AS pct
+  FROM
+    `httparchive.almanac.parsed_css`,
+    UNNEST(getVendorPrefixKeywords(css)) AS keyword
```
Comment (tunetheweb) [inline sql/2021/css/vendor_prefix_keywords.sql:97]: Potentially incorrect total

File: sql/2021/css/vendor_prefix_media.sql [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+  SELECT
+    client,
+    media,
+    COUNT(DISTINCT page) AS pages,
+    COUNT(0) AS freq,
+    SUM(COUNT(0)) OVER (PARTITION BY client) AS total,
+    COUNT(0) / SUM(COUNT(0)) OVER (PARTITION BY client) AS pct
+  FROM
+    `httparchive.almanac.parsed_css`,
+    UNNEST(getVendorPrefixMedia(css)) AS media
```
Comment (tunetheweb) [inline sql/2021/css/vendor_prefix_media.sql:95]: Potentially incorrect total

File: sql/2021/css/vendor_prefix_properties.sql [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+  SELECT
+    client,
+    prop,
+    COUNT(DISTINCT page) AS pages,
+    COUNT(0) AS freq,
+    SUM(COUNT(0)) OVER (PARTITION BY client) AS total,
+    COUNT(0) / SUM(COUNT(0)) OVER (PARTITION BY client) AS pct
+  FROM
+    `httparchive.almanac.parsed_css`,
+    UNNEST(getVendorPrefixProperties(css)) AS prop
```
Comment (tunetheweb) [inline sql/2021/css/vendor_prefix_properties.sql:98]: Potentially incorrect total

File: sql/2021/css/vendor_prefix_pseudo_classes.sql [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+  SELECT
+    client,
+    pseudo_class,
+    COUNT(DISTINCT page) AS pages,
+    COUNT(0) AS freq,
+    SUM(COUNT(0)) OVER (PARTITION BY client) AS total,
+    COUNT(0) / SUM(COUNT(0)) OVER (PARTITION BY client) AS pct
+  FROM
+    `httparchive.almanac.parsed_css`,
+    UNNEST(getVendorPrefixPseudoClasses(css)) AS pseudo_class
```
Comment (tunetheweb) [inline sql/2021/css/vendor_prefix_pseudo_classes.sql:95]: Potentially incorrect totals

File: sql/2021/css/vendor_prefix_pseudo_elements.sql [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+  SELECT
+    client,
+    pseudo_element,
+    COUNT(DISTINCT page) AS pages,
+    COUNT(0) AS freq,
+    SUM(COUNT(0)) OVER (PARTITION BY client) AS total,
+    COUNT(0) / SUM(COUNT(0)) OVER (PARTITION BY client) AS pct
+  FROM
+    `httparchive.almanac.parsed_css`,
+    UNNEST(getVendorPrefixPseudoElements(css)) AS pseudo_element
```
Comment (tunetheweb) [inline sql/2021/css/vendor_prefix_pseudo_elements.sql:95]: Potentially incorrect totals

File: sql/2021/css/vendor_prefix_summary.sql [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+  SELECT
+    client,
+    prop,
+    COUNT(DISTINCT page) AS pages,
+    COUNT(0) AS freq,
+    SUM(COUNT(IF(prop = 'total', NULL, 0))) OVER (PARTITION BY client) AS total,
+    COUNT(IF(prop = 'total', NULL, 0)) / SUM(COUNT(IF(prop = 'total', NULL, 0))) OVER (PARTITION BY client) AS pct
+  FROM
+    `httparchive.almanac.parsed_css`,
+    UNNEST(getPrefixStats(css)) AS prop
```
Comment (tunetheweb) [inline sql/2021/css/vendor_prefix_summary.sql:111]: Potentially incorrect totals

**Review by tunetheweb — COMMENTED** [review]
Pointed out some potential total issues.

File: sql/2021/css/box_sizing.sql (hunk 1 above)
Comment (rviscomi) [reply sql/2021/css/box_sizing.sql:33]: This should be ok given the use of `DISTINCT page`

File: sql/2021/css/content_values.sql (hunk 2 above)
Comment (rviscomi) [reply sql/2021/css/content_values.sql:52]: I think this is WAI: count of pages uses `DISTINCT page` and count of content items should account for each UNNESTed value.

File: sql/2021/css/custom_property_registered.sql (hunk 3 above)
Comment (rviscomi) [reply sql/2021/css/custom_property_registered.sql:51]: WAI: `DISTINCT url`

File: sql/2021/css/calc_properties.sql (hunk 5 above)
Comment (rviscomi) [reply sql/2021/css/calc_properties.sql:91]: The most significant change in `pct` values is a tenth of a percent, which isn't enough to affect the precision of the results. I agree the way this is implemented is less correct but there are many queries and they take a very long time to run, so the effort doesn't seem worth it for such small improvements.

**Review by rviscomi — COMMENTED** [review]
Unable to address all of the comments individually (GitHub is having a very hard time given all of the files) so I'll summarize here.

Some of the uses of UNNEST are working as intended. We use `DISTINCT` to deduplicate the rows accordingly and `COUNT(0)` is used as intended to count each individual row.

You have a good point in other cases about `HAVING` clauses affecting the totals. I don't necessarily think these are worth fixing, given the amount of overhead to rerun the queries compared to the small effects when including the tail data in the sums. LMK if that sits ok with you.

File: sql/2021/css/box_sizing.sql (hunk 1 above)
Comment (tunetheweb) [reply sql/2021/css/box_sizing.sql:33]: Agreed on a second look. Happy to resolve.

File: sql/2021/css/content_values.sql (hunk 2 above)
Comment (tunetheweb) [reply sql/2021/css/content_values.sql:52]: Could you double check this one? I’m less worried about the distinct page and more worried about this line (and the pct line based on it):

```sql
SUM(COUNT(0)) OVER (PARTITION BY client) AS total,
```

This has potential to be a huge increase on total number.

Unresolving for now.

File: sql/2021/css/custom_property_registered.sql (hunk 3 above)
Comment (tunetheweb) [reply sql/2021/css/custom_property_registered.sql:51]: Agree on second look. Happy to leave as resolved.

File: sql/2021/css/calc_properties.sql (hunk 5 above)
Comment (tunetheweb) [reply sql/2021/css/calc_properties.sql:91]: That’s fine as long as your happy with it. Agree impact will likely be small in most cases.

File: sql/2021/css/content_values.sql (hunk 2 above)
Comment (rviscomi) [reply sql/2021/css/content_values.sql:52]: Yeah I think it's ok. The total should be counting all `content` values. And the `WHERE pages >= 1000` is in a wrapped outer query that won't affect the total.

Comment (tunetheweb) [reply sql/2021/css/content_values.sql:52]: An you’ve a separate “total_pages” column above which is calced with a sub query, so this one isn’t a total pages. Ok that makes sense. I think :-)

**Comment by rviscomi** [comment]
@tunetheweb any outstanding concerns or is this ok to merge from your perspective?

**Review by tunetheweb — APPROVED** [review]
Heck of a job!

**Pull request merged**